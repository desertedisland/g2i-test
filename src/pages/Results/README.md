**Results Page**

This page will display the results of the quiz based on the correctAnswer / incorrectAnswer lists
in the 'score' store.

Results are displayed with 'correct' / 'incorrect' className depending on whether the answer was correct 
or not, these classNames will correspond to a unicode tick or cross in the display.
