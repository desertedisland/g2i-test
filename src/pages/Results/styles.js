import styled from 'styled-components';
import variables from '../../styles/variables';

export const AnswerDisplay = styled.ul`
      
    list-style-type: none;
    
    li{
      margin: 0;
      padding: ${variables.marginDefault / 2}px 0;
      
      &.correct:before{
        content: '✔';
        color: ${variables.primaryColor};
        padding: 8px;
        font-size: 20px;
      }

      &.incorrect:before{
        content: '✖';
        color: ${variables.dangerColor};
        padding: 8px;
        font-size: 20px;
      }
         
    
  }
  
`;
