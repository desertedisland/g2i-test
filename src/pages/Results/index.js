/* eslint react/jsx-one-expression-per-line: 0 */

import React, { useEffect } from 'react';
import { useHistory } from 'react-router-dom';
import { useSelector } from 'react-redux';
import { HeaderH1, AlignedPage, StyledLink } from '../../styles/components';
import { AnswerDisplay } from './styles';

import { htmlDecode } from '../../lib/misc'; // Decode HTML entities

export default function Results() {

  const history = useHistory();
  const score = useSelector((state) => state.score);

  useEffect(
    () => {

      // If no questions have been answered then return to the home page.
      if (!score.correctAnswers.length && !score.incorrectAnswers.length) {

        return history.push('/');

      }

    }, [],
  );

  // Create a unique key for each card by taking the first 10 characters of the
  // question, removing whitespace and transforming it to lowercase.
  const getKey = (question) => question.split('').slice(0, 10).join('')
    .replace(/\s/g, '')
    .toLowerCase();

  // Combine both correct and incorrect answers so that they can be displayed with the result
  // and correct answer.
  const results = [
    ...score.correctAnswers.map((answer) => ({ ...answer, iscorrect: true, key: getKey(answer.question) })),
    ...score.incorrectAnswers.map((answer) => ({ ...answer, iscorrect: false, key: getKey(answer.question) })),
  ];

  return (
    <AlignedPage>

      <HeaderH1 data-test-id="quiz-results-heading">
        You scored {score.correctAnswers.length} / { score.correctAnswers.length + score.incorrectAnswers.length }
      </HeaderH1>

      <AnswerDisplay data-test-id="quiz-answers-display">

        {
          results.map(
            (answer) => (
              <li key={answer.key} className={answer.iscorrect ? 'correct' : 'incorrect'}>
                { htmlDecode(answer.question) }
              </li>
            ),
          )
        }

      </AnswerDisplay>

      <StyledLink to="/">Play again?</StyledLink>

    </AlignedPage>
  );

}
