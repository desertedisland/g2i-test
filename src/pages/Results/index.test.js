/* eslint no-undef: 0 */

import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import Results from './index';

const mockStore = configureStore([]);

const testStore = {
  cards: {
    working: false,
    cards: [],
  },
  score: {
    correctAnswers: [{
      category: 'Hitchhikers guide to the galaxy',
      question: 'What is the meaning of life?',
      correct_answer: 'True',
    }],
    incorrectAnswers: [{
      category: 'Hitchhikers guide to the galaxy',
      question: 'Who is Ford Prefect?',
      correct_answer: 'True',
    }],
  },
};

const store = mockStore(testStore);

describe('Results page',
  () => {

    let testRenderer = null;

    beforeAll(
      () => {

        store.dispatch = jest.fn();
        testRenderer = renderer.create(<Provider store={store}><Results /></Provider>);

      },
    );

    test('Snapshot test:',
      () => {

        expect(testRenderer.toJSON()).toMatchSnapshot();

      });

    it('Should render the heading correctly',
      () => {

        const heading = testRenderer.root.findByProps({ 'data-test-id': 'quiz-results-heading' });
        expect(heading.props.children).toEqual(['You scored ', 1, ' / ', 2]);

      });

    it('Should render the answers correctly',
      () => {

        const answers = testRenderer.root.findByProps({ 'data-test-id': 'quiz-answers-display' });
        const answer1 = answers.props.children[0].props;
        const answer2 = answers.props.children[1].props;

        expect(answer1.className).toEqual('correct');
        expect(answer1.children).toEqual('What is the meaning of life?');

        expect(answer2.className).toEqual('incorrect');
        expect(answer2.children).toEqual('Who is Ford Prefect?');

      });

  });
