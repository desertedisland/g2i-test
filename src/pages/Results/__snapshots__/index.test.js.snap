// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Results page Snapshot test: 1`] = `
<div
  className="sc-gsTCUz fFRaSL"
>
  <h1
    className="sc-bdfBwQ CDlxa"
    data-test-id="quiz-results-heading"
  >
    You scored 
    1
     / 
    2
  </h1>
  <ul
    className="sc-jSgupP edqJPr"
    data-test-id="quiz-answers-display"
  >
    <li
      className="correct"
    >
      What is the meaning of life?
    </li>
    <li
      className="incorrect"
    >
      Who is Ford Prefect?
    </li>
  </ul>
</div>
`;
